File: web@codeforces/compete/c1200-1299/c1228/C.py
x, n = map(int, input().split())

def prime(x):
    p = 2
    P = []
    while p*p <= x:
        if x % p == 0:
            P.append(p)
            while x % p == 0 and x != 0:
                x = int(x // p)
        p += 1

    if x != 1:
        P.append(x)

    return P

def fact(n, x):
    res = 0
    while n > 0:
        res += int( n // x )
        n = int( n // x )
    return res

def expMod(b, e, m):
    if e == 0:
        return 1
    elif e == 1:
        return (b % m)
    elif e % 2 == 0:
        return expMod(b*b%m, int(e//2), m)
    elif e % 2 == 1:
        return expMod(b*b%m, int(e//2), m) * b % m

# ========================

result = 1
P = prime(x)

m = int(1e9) + 7

for p in P:
    pwr = fact(n, p)
    q = int( pwr // m )
    r = int( pwr % m )

    result *= expMod(p, q+r, m)
    result %= m

print(result)

File: web@codeforces/compete/c1200-1299/c1228/C.cpp
#include <bits/stdc++.h>
using namespace std;

long long n, x;

vector<int> prime(long long a) {
    vector<int> P;
    for(long long p = 2; p*p <= a; p++) {
        if( a % p == 0 ) {
            //cout << p << endl;
            P.push_back(p);
            while(a % p == 0) a /= p;
        }
    }
    if(a != 1) P.push_back(a);
    return P;
}

// vector<int> prime(long long a) {
//     int p = 2;
//     vector<int> P (0);
//     while(a >= p) {
//         if( a % p == 0 ) {
//             P.push_back(p);
//             while(a % p == 0 && a > 1) a /= p;
//         }
//         p++;
//     }
//     return P;
// }

long long fact(long long N, long long X) {
    long long result = 0;
    while(N > 0) {
        result += (N / X);
        N /= X;
    }
    return result;
}

long long expMod(long long b, long long e, long long M) {
    if( e == 0 )        return 1;
    if( e == 1 )        return b % M;
    if( e % 2 == 0 )    return expMod(b * b % M, e / 2, M);
    else                return expMod(b * b % M, e / 2, M) * b % M;
}

int main () {
    cin >> x >> n;
    //
    long long ans = 1;
    const long long M = (long long)(1e9) + 7;
    //
    vector<int> P = prime(x);
    //
    for(auto &p : P) {
        long long pwr = fact(n, p);

        long long q = pwr / M;
        long long r = pwr % M;

        ans *= expMod(p, q+r, M);
        ans %= M;
    }
    //
    cout << ans;
}